File: src/metrics.rs
use crate::model::{
    CharStats, KeyAction, KeyEventRecord, PracticeTarget, SessionRecord, TokenKind, TokenStat,
};
use chrono::{DateTime, Utc};
use std::collections::BTreeMap;
use uuid::Uuid;

#[derive(Debug, Clone)]
pub struct TokenSpan {
    pub token: String,
    pub kind: TokenKind,
    pub start: usize,
    pub end: usize,
}

const SYMBOL_PATTERNS: &[&str] = &[
    "!==", "===", "=>", "&&", "||", ">=", "<=", "?.", "??", "${}", "()", "[]", "{}", "<>", "''",
    "\"\"", "``", "::", "->", "+=", "-=", "*=", "/=", "_", "-", "=", "+", "*", "/", "\\", "?", "!",
    ":", ";", ",", ".", "(", ")", "[", "]", "{", "}", "<", ">", "'", "\"", "`",
];
const IDLE_THRESHOLD_MS: u64 = 10_000;

#[derive(Debug, Clone, Copy, Default)]
struct TimingBreakdown {
    active_ms: u64,
    idle_ms: u64,
    idle_pause_count: u32,
    start_to_first_key_ms: u64,
    last_key_to_end_ms: u64,
}

pub fn build_session_record(
    target: PracticeTarget,
    started_at: DateTime<Utc>,
    duration_ms: u64,
    manual_pause_ms: u64,
    user_input: String,
    key_events: Vec<KeyEventRecord>,
) -> SessionRecord {
    let target_chars: Vec<char> = target.text.chars().collect();
    let input_chars: Vec<char> = user_input.chars().collect();
    let target_len = target_chars.len();
    let insert_count = key_events
        .iter()
        .filter(|event| matches!(event.action, KeyAction::Insert))
        .count();
    let correct_insert_count = key_events
        .iter()
        .filter(|event| matches!(event.action, KeyAction::Insert) && event.correct)
        .count();
    let has_auto_indent = key_events
        .iter()
        .any(|event| matches!(event.action, KeyAction::AutoIndent));
    let backspace_count = key_events
        .iter()
        .filter(|event| matches!(event.action, KeyAction::Backspace))
        .count() as u32;
    let final_correct_chars = target_chars
        .iter()
        .zip(input_chars.iter())
        .filter(|(expected, actual)| expected == actual)
        .count();
    let correct_chars = if has_auto_indent {
        correct_insert_count
    } else {
        final_correct_chars
    };

    let mut error_chars = BTreeMap::<String, u32>::new();
    let error_count = key_events
        .iter()
        .filter(|event| matches!(event.action, KeyAction::Insert) && !event.correct)
        .inspect(|event| {
            let key = event
                .expected
                .or(event.input)
                .map(|ch| printable_char(ch).to_string())
                .unwrap_or_else(|| "<extra>".into());
            *error_chars.entry(key).or_default() += 1;
        })
        .count() as u32;
    let timing = timing_breakdown(duration_ms, &key_events);
    let adjusted_key_events = adjusted_key_events(&key_events, timing.start_to_first_key_ms);

    let minutes = (timing.active_ms.max(1) as f64) / 60_000.0;
    let raw_wpm = insert_count as f64 / 5.0 / minutes;
    let wpm = correct_chars as f64 / 5.0 / minutes;
    let accuracy = if insert_count == 0 {
        0.0
    } else {
        correct_insert_count as f64 / insert_count as f64 * 100.0
    };

    let token_stats = collect_token_stats(&target.text, &adjusted_key_events);
    let mut error_tokens = BTreeMap::<String, u32>::new();
    for stat in &token_stats {
        if stat.errors > 0 {
            *error_tokens.entry(stat.token.clone()).or_default() += stat.errors;
        }
    }

    let mut slow_tokens = token_stats.clone();
    slow_tokens.sort_by_key(|stat| {
        std::cmp::Reverse(stat.start_delay_ms + stat.duration_ms / 2 + u64::from(stat.errors) * 250)
    });
    slow_tokens.truncate(12);

    SessionRecord {
        id: Uuid::new_v4().to_string(),
        started_at,
        mode: target.mode,
        source: target.source,
        daily_run_id: String::new(),
        lesson_id: String::new(),
        lesson_index: None,
        completion_state: Default::default(),
        module: Default::default(),
        category: Default::default(),
        active_ms: timing.active_ms,
        idle_ms: timing.idle_ms,
        manual_pause_ms,
        idle_pause_count: timing.idle_pause_count,
        start_to_first_key_ms: timing.start_to_first_key_ms,
        last_key_to_end_ms: timing.last_key_to_end_ms,
        char_stats: CharStats {
            correct: correct_chars,
            incorrect: error_count as usize,
            extra: input_chars.len().saturating_sub(target_chars.len()),
            missed: target_len.saturating_sub(input_chars.len()),
        },
        duration_ms,
        target_text: target.text,
        user_input,
        target_len,
        typed_len: insert_count,
        correct_chars,
        wpm,
        raw_wpm,
        accuracy,
        error_count,
        backspace_count,
        error_chars,
        error_tokens,
        slow_tokens,
        token_stats,
        key_events,
    }
}

fn timing_breakdown(duration_ms: u64, key_events: &[KeyEventRecord]) -> TimingBreakdown {
    let Some(first) = key_events.first() else {
        return TimingBreakdown::default();
    };
    let last = key_events.last().expect("first event exists");
    let start_to_first_key_ms = first.at_ms.min(duration_ms);
    let last_key_to_end_ms = duration_ms.saturating_sub(last.at_ms.min(duration_ms));
    let mut idle_ms = 0;
    let mut idle_pause_count = 0;
    for pair in key_events.windows(2) {
        let gap = pair[1].at_ms.saturating_sub(pair[0].at_ms);
        if gap > IDLE_THRESHOLD_MS {
            idle_pause_count += 1;
            idle_ms += gap - IDLE_THRESHOLD_MS;
        }
    }
    let active_ms = duration_ms
        .saturating_sub(start_to_first_key_ms)
        .saturating_sub(last_key_to_end_ms)
        .saturating_sub(idle_ms);

    TimingBreakdown {
        active_ms,
        idle_ms,
        idle_pause_count,
        start_to_first_key_ms,
        last_key_to_end_ms,
    }
}

fn adjusted_key_events(
    key_events: &[KeyEventRecord],
    start_to_first_key_ms: u64,
) -> Vec<KeyEventRecord> {
    let mut adjusted = Vec::with_capacity(key_events.len());
    let mut idle_excess_before = 0;
    let mut previous_at = None;
    for event in key_events {
        if let Some(previous_at) = previous_at {
            let gap = event.at_ms.saturating_sub(previous_at);
            if gap > IDLE_THRESHOLD_MS {
                idle_excess_before += gap - IDLE_THRESHOLD_MS;
            }
        }
        let mut event = event.clone();
        event.at_ms = event
            .at_ms
            .saturating_sub(start_to_first_key_ms)
            .saturating_sub(idle_excess_before);
        previous_at = Some(event.at_ms + start_to_first_key_ms + idle_excess_before);
        adjusted.push(event);
    }
    adjusted
}

pub fn token_spans(text: &str) -> Vec<TokenSpan> {
    let chars: Vec<char> = text.chars().collect();
    let mut spans = Vec::new();
    let mut index = 0;

    while index < chars.len() {
        if chars[index].is_whitespace() {
            index += 1;
            continue;
        }

        if is_word_start(chars[index]) {
            let start = index;
            index += 1;
            while index < chars.len() && is_word_continue(chars[index]) {
                index += 1;
            }
            let token: String = chars[start..index].iter().collect();
            let kind = if token == "_" {
                TokenKind::Symbol
            } else {
                TokenKind::Word
            };
            spans.push(TokenSpan {
                token,
                kind,
                start,
                end: index,
            });
            continue;
        }

        if let Some(pattern) = match_symbol_at(&chars, index) {
            let len = pattern.chars().count();
            spans.push(TokenSpan {
                token: pattern.to_string(),
                kind: TokenKind::Symbol,
                start: index,
                end: index + len,
            });
            index += len;
            continue;
        }

        spans.push(TokenSpan {
            token: chars[index].to_string(),
            kind: TokenKind::Code,
            start: index,
            end: index + 1,
        });
        index += 1;
    }

    spans
}

fn collect_token_stats(text: &str, key_events: &[KeyEventRecord]) -> Vec<TokenStat> {
    let spans = token_spans(text);
    let inserts: Vec<&KeyEventRecord> = key_events
        .iter()
        .filter(|event| matches!(event.action, KeyAction::Insert))
        .collect();

    let mut visible_at_ms = 0;
    let mut stats = Vec::new();

    for span in spans {
        let events: Vec<&KeyEventRecord> = inserts
            .iter()
            .copied()
            .filter(|event| {
                event.at_ms >= visible_at_ms
                    && event.position >= span.start
                    && event.position < span.end
            })
            .collect();

        if let (Some(first), Some(last)) = (events.first(), events.last()) {
            let errors = events.iter().filter(|event| !event.correct).count() as u32;
            stats.push(TokenStat {
                token: span.token,
                kind: span.kind,
                start_delay_ms: first.at_ms.saturating_sub(visible_at_ms),
                duration_ms: last.at_ms.saturating_sub(first.at_ms),
                errors,
            });
            visible_at_ms = last.at_ms;
        }
    }

    stats
}

fn match_symbol_at(chars: &[char], index: usize) -> Option<&'static str> {
    SYMBOL_PATTERNS.iter().copied().find(|pattern| {
        let pattern_chars: Vec<char> = pattern.chars().collect();
        let end = index + pattern_chars.len();
        end <= chars.len() && chars[index..end] == pattern_chars
    })
}

fn is_word_start(ch: char) -> bool {
    ch.is_ascii_alphabetic() || ch == '_'
}

fn is_word_continue(ch: char) -> bool {
    ch.is_ascii_alphanumeric() || ch == '_'
}

fn printable_char(ch: char) -> String {
    match ch {
        '\n' => "\\n".into(),
        '\t' => "\\t".into(),
        ' ' => "<space>".into(),
        other => other.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tokenizes_words_and_programming_symbols() {
        let tokens: Vec<String> = token_spans("items.map((item) => item.id !== null)")
            .into_iter()
            .map(|span| span.token)
            .collect();

        assert!(tokens.contains(&"items".to_string()));
        assert!(tokens.contains(&"=>".to_string()));
        assert!(tokens.contains(&"!==".to_string()));
    }

    #[test]
    fn accuracy_counts_corrected_mistakes() {
        let target = PracticeTarget {
            mode: crate::model::Mode::Words,
            text: "abc".to_string(),
            source: "test".to_string(),
        };
        let events = vec![
            KeyEventRecord {
                at_ms: 100,
                action: KeyAction::Insert,
                position: 0,
                expected: Some('a'),
                input: Some('a'),
                correct: true,
            },
            KeyEventRecord {
                at_ms: 200,
                action: KeyAction::Insert,
                position: 1,
                expected: Some('b'),
                input: Some('x'),
                correct: false,
            },
            KeyEventRecord {
                at_ms: 300,
                action: KeyAction::Backspace,
                position: 1,
                expected: Some('b'),
                input: None,
                correct: false,
            },
            KeyEventRecord {
                at_ms: 400,
                action: KeyAction::Insert,
                position: 1,
                expected: Some('b'),
                input: Some('b'),
                correct: true,
            },
            KeyEventRecord {
                at_ms: 500,
                action: KeyAction::Insert,
                position: 2,
                expected: Some('c'),
                input: Some('c'),
                correct: true,
            },
        ];

        let record = build_session_record(target, Utc::now(), 1000, 0, "abc".to_string(), events);

        assert_eq!(record.correct_chars, 3);
        assert_eq!(record.typed_len, 4);
        assert_eq!(record.error_count, 1);
        assert_eq!(record.accuracy, 75.0);
    }

    #[test]
    fn wpm_excludes_start_delay_and_last_key_tail() {
        let target = PracticeTarget {
            mode: crate::model::Mode::Words,
            text: "abc".to_string(),
            source: "test".to_string(),
        };
        let events = vec![
            insert(5_000, 0, 'a', true),
            insert(5_500, 1, 'b', true),
            insert(6_000, 2, 'c', true),
        ];

        let record = build_session_record(target, Utc::now(), 20_000, 0, "abc".to_string(), events);

        assert_eq!(record.start_to_first_key_ms, 5_000);
        assert_eq!(record.last_key_to_end_ms, 14_000);
        assert_eq!(record.active_ms, 1_000);
        assert!((record.wpm - 36.0).abs() < f64::EPSILON);
    }

    #[test]
    fn idle_gap_excess_is_excluded_from_wpm_and_token_stats() {
        let target = PracticeTarget {
            mode: crate::model::Mode::Words,
            text: "ab cd".to_string(),
            source: "test".to_string(),
        };
        let events = vec![
            insert(100, 0, 'a', true),
            insert(200, 1, 'b', true),
            insert(20_300, 3, 'c', true),
            insert(20_400, 4, 'd', true),
        ];

        let record =
            build_session_record(target, Utc::now(), 20_500, 0, "ab cd".to_string(), events);

        assert_eq!(record.idle_pause_count, 1);
        assert_eq!(record.idle_ms, 10_100);
        assert_eq!(record.active_ms, 10_200);
        let cd = record
            .token_stats
            .iter()
            .find(|stat| stat.token == "cd")
            .expect("cd token should be measured");
        assert_eq!(cd.start_delay_ms, 10_000);
    }

    #[test]
    fn token_stats_ignore_obsolete_events_after_cross_token_retype() {
        let events = vec![
            insert(100, 0, 'f', true),
            insert(200, 1, 'o', true),
            insert(300, 2, 'o', true),
            insert(400, 4, 'b', true),
            insert(500, 5, 'a', true),
            insert(600, 6, 'x', false),
            insert(900, 2, 'o', true),
            insert(1000, 4, 'b', true),
            insert(1100, 5, 'a', true),
            insert(1200, 6, 'r', true),
        ];

        let stats = collect_token_stats("foo bar", &events);
        let bar = stats
            .iter()
            .find(|stat| stat.token == "bar")
            .expect("bar token should be measured");

        assert_eq!(bar.start_delay_ms, 100);
        assert_eq!(bar.duration_ms, 200);
        assert_eq!(bar.errors, 0);
    }

    fn insert(at_ms: u64, position: usize, input: char, correct: bool) -> KeyEventRecord {
        KeyEventRecord {
            at_ms,
            action: KeyAction::Insert,
            position,
            expected: Some(input),
            input: Some(input),
            correct,
        }
    }
}

File: src/training/metrics.ts
import { randomUUID } from "node:crypto";

import type {
  KeyEventRecord,
  PracticeTarget,
  SessionRecord,
  TokenKind,
  TokenStat,
} from "../domain/model";

export interface TokenSpan {
  token: string;
  kind: TokenKind;
  start: number;
  end: number;
}

interface TimingBreakdown {
  active_ms: number;
  idle_ms: number;
  idle_pause_count: number;
  start_to_first_key_ms: number;
  last_key_to_end_ms: number;
}

const SYMBOL_PATTERNS = [
  "!==",
  "===",
  "=>",
  "&&",
  "||",
  ">=",
  "<=",
  "?.",
  "??",
  "${}",
  "()",
  "[]",
  "{}",
  "<>",
  "''",
  '""',
  "``",
  "::",
  "->",
  "+=",
  "-=",
  "*=",
  "/=",
  "_",
  "-",
  "=",
  "+",
  "*",
  "/",
  "\\",
  "?",
  "!",
  ":",
  ";",
  ",",
  ".",
  "(",
  ")",
  "[",
  "]",
  "{",
  "}",
  "<",
  ">",
  "'",
  '"',
  "`",
] as const;

// 按首字符索引，避免每个字符位置都线性扫描全部 48 个 pattern。
// 组内保持 SYMBOL_PATTERNS 的原始顺序（长度降序），从而维持"最长优先"匹配语义。
const SYMBOL_PATTERNS_BY_FIRST_CHAR: ReadonlyMap<string, readonly string[]> = (() => {
  const map = new Map<string, string[]>();
  for (const pattern of SYMBOL_PATTERNS) {
    const first = Array.from(pattern)[0];
    if (first === undefined) {
      continue;
    }
    const group = map.get(first) ?? [];
    group.push(pattern);
    map.set(first, group);
  }
  return map;
})();

const IDLE_THRESHOLD_MS = 10_000;

export function buildSessionRecord(
  target: PracticeTarget,
  started_at: string,
  duration_ms: number,
  manual_pause_ms: number,
  user_input: string,
  key_events: KeyEventRecord[],
): SessionRecord {
  const targetChars = Array.from(target.text);
  const inputChars = Array.from(user_input);
  const targetLen = targetChars.length;
  const insertEvents = key_events.filter((event) => event.action === "insert");
  const insertCount = insertEvents.length;
  const correctInsertCount = insertEvents.filter((event) => event.correct).length;
  const hasAutoIndent = key_events.some((event) => event.action === "auto_indent");
  const backspaceCount = key_events.filter((event) => event.action === "backspace").length;
  const finalCorrectChars = targetChars.filter(
    (expected, index) => inputChars[index] === expected,
  ).length;
  const correctChars = hasAutoIndent ? correctInsertCount : finalCorrectChars;

  const errorChars: Record<string, number> = {};
  let errorCount = 0;
  for (const event of insertEvents) {
    if (event.correct) {
      continue;
    }
    errorCount += 1;
    const key = event.expected ?? event.input;
    const label = key === null ? "<extra>" : printableChar(key);
    errorChars[label] = (errorChars[label] ?? 0) + 1;
  }

  const timing = timingBreakdown(duration_ms, key_events);
  const normalizedKeyEvents = adjustedKeyEvents(
    key_events,
    timing.start_to_first_key_ms,
  );
  const minutes = Math.max(timing.active_ms, 1) / 60_000;
  const rawWpm = insertCount / 5 / minutes;
  const wpm = correctChars / 5 / minutes;
  const accuracy = insertCount === 0 ? 0 : (correctInsertCount / insertCount) * 100;
  const tokenStats = collectTokenStats(target.text, normalizedKeyEvents);
  const errorTokens: Record<string, number> = {};

  for (const stat of tokenStats) {
    if (stat.errors > 0) {
      errorTokens[stat.token] = (errorTokens[stat.token] ?? 0) + stat.errors;
    }
  }

  const slowTokens = [...tokenStats]
    .sort(
      (left, right) =>
        tokenScore(right) - tokenScore(left),
    )
    .slice(0, 12);

  return {
    id: randomUUID(),
    started_at,
    mode: target.mode,
    source: target.source,
    daily_run_id: "",
    lesson_id: "",
    lesson_index: null,
    completion_state: "completed",
    module: "unknown",
    category: "unknown",
    active_ms: timing.active_ms,
    idle_ms: timing.idle_ms,
    manual_pause_ms,
    idle_pause_count: timing.idle_pause_count,
    start_to_first_key_ms: timing.start_to_first_key_ms,
    last_key_to_end_ms: timing.last_key_to_end_ms,
    char_stats: {
      correct: correctChars,
      incorrect: errorCount,
      extra: saturatingSub(inputChars.length, targetChars.length),
      missed: saturatingSub(targetLen, inputChars.length),
    },
    duration_ms,
    target_text: target.text,
    user_input,
    target_len: targetLen,
    typed_len: insertCount,
    correct_chars: correctChars,
    wpm,
    raw_wpm: rawWpm,
    accuracy,
    error_count: errorCount,
    backspace_count: backspaceCount,
    error_chars: errorChars,
    error_tokens: errorTokens,
    slow_tokens: slowTokens,
    token_stats: tokenStats,
    key_events,
  };
}

export function tokenSpans(text: string): TokenSpan[] {
  const chars = Array.from(text);
  const spans: TokenSpan[] = [];
  let index = 0;

  while (index < chars.length) {
    const current = chars[index];
    if (current === undefined) {
      break;
    }

    if (isWhitespace(current)) {
      index += 1;
      continue;
    }

    if (isWordStart(current)) {
      const start = index;
      index += 1;
      while (index < chars.length) {
        const next = chars[index];
        if (next === undefined || !isWordContinue(next)) {
          break;
        }
        index += 1;
      }
      const token = chars.slice(start, index).join("");
      spans.push({
        token,
        kind: token === "_" ? "symbol" : "word",
        start,
        end: index,
      });
      continue;
    }

    const pattern = matchSymbolAt(chars, index);
    if (pattern !== null) {
      const length = Array.from(pattern).length;
      spans.push({
        token: pattern,
        kind: "symbol",
        start: index,
        end: index + length,
      });
      index += length;
      continue;
    }

    spans.push({
      token: current,
      kind: "code",
      start: index,
      end: index + 1,
    });
    index += 1;
  }

  return spans;
}

function collectTokenStats(
  text: string,
  keyEvents: KeyEventRecord[],
): TokenStat[] {
  const spans = tokenSpans(text);
  const inserts = keyEvents.filter((event) => event.action === "insert");
  let visibleAtMs = 0;
  const stats: TokenStat[] = [];

  for (const span of spans) {
    const events = inserts.filter(
      (event) =>
        event.at_ms >= visibleAtMs &&
        event.position >= span.start &&
        event.position < span.end,
    );
    const first = events[0];
    const last = events.at(-1);
    if (first === undefined || last === undefined) {
      continue;
    }

    stats.push({
      token: span.token,
      kind: span.kind,
      start_delay_ms: saturatingSub(first.at_ms, visibleAtMs),
      duration_ms: saturatingSub(last.at_ms, first.at_ms),
      errors: events.filter((event) => !event.correct).length,
    });
    visibleAtMs = last.at_ms;
  }

  return stats;
}

function timingBreakdown(
  durationMs: number,
  keyEvents: KeyEventRecord[],
): TimingBreakdown {
  const first = keyEvents[0];
  if (first === undefined) {
    return {
      active_ms: 0,
      idle_ms: 0,
      idle_pause_count: 0,
      start_to_first_key_ms: 0,
      last_key_to_end_ms: 0,
    };
  }

  const last = keyEvents.at(-1);
  if (last === undefined) {
    throw new Error("unreachable: first key event exists");
  }

  const startToFirstKeyMs = Math.min(first.at_ms, durationMs);
  const lastKeyToEndMs = saturatingSub(durationMs, Math.min(last.at_ms, durationMs));
  let idleMs = 0;
  let idlePauseCount = 0;

  for (let index = 1; index < keyEvents.length; index += 1) {
    const previous = keyEvents[index - 1];
    const current = keyEvents[index];
    if (previous === undefined || current === undefined) {
      continue;
    }
    const gap = saturatingSub(current.at_ms, previous.at_ms);
    if (gap > IDLE_THRESHOLD_MS) {
      idlePauseCount += 1;
      idleMs += gap - IDLE_THRESHOLD_MS;
    }
  }

  return {
    active_ms: saturatingSub(
      saturatingSub(
        saturatingSub(durationMs, startToFirstKeyMs),
        lastKeyToEndMs,
      ),
      idleMs,
    ),
    idle_ms: idleMs,
    idle_pause_count: idlePauseCount,
    start_to_first_key_ms: startToFirstKeyMs,
    last_key_to_end_ms: lastKeyToEndMs,
  };
}

function adjustedKeyEvents(
  keyEvents: KeyEventRecord[],
  startToFirstKeyMs: number,
): KeyEventRecord[] {
  const adjusted: KeyEventRecord[] = [];
  let idleExcessBefore = 0;
  let previousAt: number | null = null;

  for (const event of keyEvents) {
    if (previousAt !== null) {
      const gap = saturatingSub(event.at_ms, previousAt);
      if (gap > IDLE_THRESHOLD_MS) {
        idleExcessBefore += gap - IDLE_THRESHOLD_MS;
      }
    }

    const atMs = saturatingSub(
      saturatingSub(event.at_ms, startToFirstKeyMs),
      idleExcessBefore,
    );
    previousAt = atMs + startToFirstKeyMs + idleExcessBefore;
    adjusted.push({ ...event, at_ms: atMs });
  }

  return adjusted;
}

function matchSymbolAt(chars: string[], index: number): string | null {
  const current = chars[index];
  if (current === undefined) {
    return null;
  }
  const candidates = SYMBOL_PATTERNS_BY_FIRST_CHAR.get(current);
  if (candidates === undefined) {
    return null;
  }
  for (const pattern of candidates) {
    const patternChars = Array.from(pattern);
    const end = index + patternChars.length;
    if (end <= chars.length && chars.slice(index, end).join("") === pattern) {
      return pattern;
    }
  }
  return null;
}

function tokenScore(stat: TokenStat): number {
  return stat.start_delay_ms + stat.duration_ms / 2 + stat.errors * 250;
}

function isWhitespace(value: string): boolean {
  return /\s/u.test(value);
}

function isWordStart(value: string): boolean {
  return /^[A-Za-z_]$/u.test(value);
}

function isWordContinue(value: string): boolean {
  return /^[A-Za-z0-9_]$/u.test(value);
}

function printableChar(value: string): string {
  switch (value) {
    case "\n":
      return "\\n";
    case "\t":
      return "\\t";
    case " ":
      return "<space>";
    default:
      return value;
  }
}

function saturatingSub(left: number, right: number): number {
  return Math.max(0, left - right);
}
